### PR TITLE
DB-9012 Flatten half outer joins

### PR DESCRIPTION
**Pull request by yxia92**

### CODE REVIEW

**Comment by yxia92** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/JoinNode.java [hunk 1]
```diff
@@ -873,6 +873,7 @@ public void pushExpressions(PredicateList outerPredicateList) throws StandardExc
      * @param sql          The SubqueryList from the outer query
      * @param gbl          The group by list, if any
      * @param havingClause The HAVING clause, if any
+     * @param numTables     maximum number of tables in the query
```
Comment (arnaud-splice) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/JoinNode.java]: rename to maxNumTables

File: hbase_sql/src/test/java/com/splicemachine/hbase/CostEstimationIT.java [hunk 2]
```diff
@@ -225,23 +225,25 @@ public void testOuterJoinRowCount() throws Exception {
...
+                    ->  ProjectRestrict(n=6,totalCost=21.497,outputRows=219,outputHeapSize=332 B,partitions=1)
+                      ->  BroadcastLeftOuterJoin(n=5,totalCost=21.497,outputRows=219,outputHeapSize=332 B,partitions=1,preds=[(A2[8:2] = A3[8:3])])
+                        ->  TableScan[T33(48224)](n=4,totalCost=4.04,scannedRows=20,outputRows=20,outputHeapSize=332 B,partitions=1)
+                        ->  BroadcastJoin(n=3,totalCost=13.039,outputRows=219,outputHeapSize=272 B,partitions=1,preds=[(A1[4:1] = A2[4:2])])
+                          ->  TableScan[T22(48208)](n=2,totalCost=4.04,scannedRows=20,outputRows=18,outputHeapSize=272 B,partitions=1,preds=[(A2[2:1] = 90)])
+                          ->  TableScan[T11(48192)](n=1,totalCost=4.6,scannedRows=300,outputRows=270,outputHeapSize=270 B,partitions=1,preds=[(A1[0:1] = 90)])
+
+            10 rows selected
 
             10 rows selected
```
Comment (arnaud-splice) [inline hbase_sql/src/test/java/com/splicemachine/hbase/CostEstimationIT.java]: remove duplicated line

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/FromTable.java [hunk 3]
```diff
@@ -123,6 +123,12 @@
      * SSQ join related variables
      */
     public boolean fromSSQ;
+
+    /* variable tracking the info of a FromTable node flattened from a HalfJoinNode.
+     * if OJLevel = 0, the table is free to join with other tables in the FromList, if OJLevel > 0, it can only
+     * join with its left table indicated in the dependencyMap */
+    protected int OJLevel;
```
[the remark this answers is not in the document]
Comment (yxia92) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/FromTable.java]: Done!

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/Predicate.java [hunk 4]
```diff
@@ -67,6 +67,8 @@
     protected boolean isQualifier;
     // Whether this is full outer join predicate
     protected boolean  forFullJoin;
+    // indicate whether the predicate is from the ON cluase of a particular half outer join
+    protected int OJLevel;
```
[the remark this answers is not in the document]
Comment (yxia92) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/Predicate.java]: I renamed it to outerJoinLevel.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/FromList.java [hunk 5]
```diff
@@ -700,6 +703,10 @@ void pushPredicates(PredicateList predicateList) throws StandardException{
         int size=size();
         for(int index=0;index<size;index++){
             FromTable fromTable=(FromTable)elementAt(index);
+            // to stay safe, we will not push where clause condition(including flattened outer join
```
[the remark this answers is not in the document]
Comment (yxia92) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/FromList.java]: This is just to simplify the logic and reduce the risk, we can enhance it in the future. This is consistent with the existing behavior. I rephrased the comment.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/GenericStatement.java [hunk 6]
```diff
@@ -564,6 +564,19 @@ else if (nativeSparkAggregationModeString.equals("forced"))
             }
             cc.setCurrentTimestampPrecision(currentTimestampPrecision);
 
+            String outerJoinFlatteningDisabledString =
+                    PropertyUtil.getCachedDatabaseProperty(lcc, Property.OUTERJOIN_FLATTENING_DISABLED);
+            boolean outerJoinFatteningDisabled = CompilerContext.DEFAULT_OUTERJOIN_FLATTENING_DISABLED;
```
[the remark this answers is not in the document]
Comment (yxia92) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/GenericStatement.java]: I've corrected the typo, thanks!

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/CompilerContextImpl.java [hunk 7]
```diff
@@ -141,6 +141,8 @@ public	void	resetContext()
         selectivityEstimationIncludingSkewedDefault = false;
         projectionPruningEnabled = false;
         maxMulticolumnProbeValues = DEFAULT_MAX_MULTICOLUMN_PROBE_VALUES;
+        nextOJLevel = 1;
+		outerJoinFlatteningDisabled = false;
```
[the remark this answers is not in the document]
Comment (yxia92) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/CompilerContextImpl.java]: Done.

File: hbase_sql/src/test/java/com/splicemachine/hbase/CostEstimationIT.java (hunk 2 above)
Comment (yxia92) [reply hbase_sql/src/test/java/com/splicemachine/hbase/CostEstimationIT.java]: Done!

**Pull request merged**